# Quarterly Sales Tax Revenue Analysis

### [Data Source](https://tax.utah.gov/econstats/sales)

### Q3: October, November, December

### Notes:
- Q3 2020 excel sheet does not contain totals for the 'UNKNOWN/NONCLASSIFIABLE' category; respective column for 2019 was removed

In [1]:
import pandas as pd
import os
import numpy as np
import arcpy
from arcgis.features import GeoAccessor, GeoSeriesAccessor
arcpy.env.overwriteOutput = True

In [2]:
# show all columns
pd.options.display.max_columns = None

In [3]:
# Removes misc characters from string
def replace(string):
    string = string.replace("&", "")
    string = string.replace("-", " ")
    string = string.replace("(", "")
    string = string.replace(")", "")
    string = string.replace("/", " ")
    string = string.replace(",", "")
    string = string.replace(".", "")
    string = string.replace(" ", "_")
    return string


### Load Q3 2020 sales tax revenue data

In [4]:
# load excel sheet into a dataframe
xlsx = '.\\Inputs\\2020-03-quarterly-sales.xlsx'
q3_2020 = pd.read_excel(xlsx, sheet_name='Table 9', header=5)

# Drop last empty two rows
q3_2020.drop(q3_2020.tail(2).index,inplace=True)

# forward fill values from merged cells
q3_2020['County'] = q3_2020['County'].fillna(method='ffill')
q3_2020['Location Code'] = q3_2020['Location Code'].fillna(method='ffill')
q3_2020['City'] = q3_2020['City'].fillna(method='ffill')

# Figures with less than 10 taxpayers have been rounded up per Tax Commission disclosure rules.
# Rename column and convert values to boolean
q3_2020.rename(columns={'Unnamed: 5':'Rounded Up'}, inplace=True)
crosswalk = {'*':True, np.nan:False} # County_ID
q3_2020['Rounded Up'] = q3_2020['Rounded Up'].map(crosswalk)

q3_2020.head()


County  Location Code         City  \
0  Beaver County         1002.0  Beaver City   
1  Beaver County         1002.0  Beaver City   
2  Beaver County         1002.0  Beaver City   
3  Beaver County         1002.0  Beaver City   
4  Beaver County         1002.0  Beaver City   

                    Economic Sector (NAICS Code)     2020Q3  Rounded Up  
0  AGRICULTURE, FORESTRY, FISHING & HUNTING (11)    80000.0        True  
1                                 UTILITIES (22)   700000.0        True  
2                              CONSTRUCTION (23)   150000.0        True  
3                          MANUFACTURING (31-33)  1880967.0       False  
4            WHOLESALE TRADE-DURABLE GOODS (423)   310250.0       False

In [5]:
# pivot the table to get sectors as columns with tax values by city
sectors_q3_2020 = pd.pivot_table(q3_2020,values = '2020Q3',index ='City', columns = 'Economic Sector (NAICS Code)', 
                                 aggfunc='first')

# fill NAs with 0
sectors_q3_2020 = sectors_q3_2020.fillna(0)

In [6]:
sectors_q3_2020.columns

Index(['ACCOMMODATION (721)',
       'ADMIN. & SUPPORT & WASTE MANAG. & REMED. SERVICES (56)',
       'AGRICULTURE, FORESTRY, FISHING & HUNTING (11)',
       'ARTS, ENTERTAINMENT AND RECREATION (71)', 'CONSTRUCTION (23)',
       'EDUCATIONAL SERVICES (61)', 'FINANCE & INSURANCE (52)',
       'FOOD SERVICES & DRINKING PLACES (722)',
       'HEALTH CARE & SOCIAL ASSISTANCE (62)', 'INFORMATION (51)',
       'MANAGEMENT OF COMPANIES & ENTERPRISES (55)', 'MANUFACTURING (31-33)',
       'MINING, QUARRYING, & OIL & GAS EXTRACTION (21)',
       'OTHER SERVICES-EXCEPT PUBLIC ADMINISTRATION (81)',
       'PRIOR-PERIOD PAYMENTS & REFUNDS', 'PRIVATE MOTOR VEHICLE SALES',
       'PROFESSIONAL, SCIENTIFIC & TECHNICAL SERVICES (54)',
       'PUBLIC ADMINISTRATION (92)', 'REAL ESTATE, RENTAL & LEASING (53)',
       'RETAIL-BUILD. MATERIAL, GARDEN EQUIP. & SUPPLIES DEALERS (444)',
       'RETAIL-CLOTHING & CLOTHING ACCESSORIES STORES (448)',
       'RETAIL-ELECTRONICS & APPLIANCE STORES (443)',
       

In [7]:
# fix column names
#new_names = [replace(item) for item in list(sectors_q3_2020.columns)]
new_names = ['L_ACCOMMODATION_20', 'S_ADMIN_SUPPORT_20','A_AG_WILDLIFE_20','L_CULTURAL_REC_20','C_CONSTRUCTION_20',
'E_EDUCATION_20','S_FINANCIAL_20','L_RSTRNT_BAR_20','H_HEALTH_CARE_20','S_IT_DATA_20','S_CORPORATE_MGMT_20',
'M_MANUFACTURING_20','E_EXTRACTION_20','O_OTHER_20','X_ADJUSTMENTS_20','R_AUTO_PRIVATE_20','S_PROF_TECH_SRV_20',
'G_GOVERNMENT_20','S_RENTAL_LEASING_20','R_BUILDING_SUPPLY_20','R_CLOTHING_20','R_ELECTRONICS_20','R_GROCERY_BEV_20',
'R_FURNITURE_20','R_GAS_STATIONS_20','R_GENERAL_RETAIL_20','R_HEALTH_RETAIL_20','R_OTHER_RETAIL_20','R_AUTO_RETAIL_20',
'R_NONSTORE_RETAIL_20','R_SPORT_HOBBY_20','L_SPECIAL_EVENT_20','W_DISTRIBUTION_20','U_UTILITIES_20',
'W_WHLSALE_DURABLE_20','S_WHLSLE_ETRADE_20','W_WHLSLE_NDURABLE_20']
sectors_q3_2020.columns = new_names

# get the total sales tax revenue
sectors_q3_2020['TOTAL_20'] = sectors_q3_2020[new_names].sum(axis=1)

sectors_q3_2020.head()

L_ACCOMMODATION_20  S_ADMIN_SUPPORT_20  A_AG_WILDLIFE_20  \
City                                                                      
American Fork            610610.0            684390.0          700000.0   
Beaver City             1796700.0             22137.0           80000.0   
Blanding                 641588.0             63792.0            1000.0   
Bountiful                200000.0           1528237.0            2971.0   
Brigham                 1237268.0            198096.0            3000.0   

               L_CULTURAL_REC_20  C_CONSTRUCTION_20  E_EDUCATION_20  \
City                                                                  
American Fork          1742593.0          9050703.0        196272.0   
Beaver City               7000.0           150000.0          6000.0   
Blanding                 40000.0           150000.0        150000.0   
Bountiful              1098401.0          1976421.0        143845.0   
Brigham                 700000.0           611762.0         54266.0   

               S_FINANCIAL_20  L_RSTRNT_BAR_20  H_HEALTH_CARE_20  \
City                                                               
American Fork       2320119.0       27458710.0          409886.0   
Beaver City          487846.0        3207140.0           60000.0   
Blanding              18431.0         450000.0           40000.0   
Bountiful           1263042.0       13021850.0         1229667.0   
Brigham              293778.0        7589533.0          349797.0   

               S_IT_DATA_20  S_CORPORATE_MGMT_20  M_MANUFACTURING_20  \
City                                                                   
American Fork     7291812.0               2000.0           3371140.0   
Beaver City        470173.0               1000.0           1880967.0   
Blanding           454379.0               3000.0            128084.0   
Bountiful         5588283.0              25000.0           2579458.0   
Brigham           2658038.0               8000.0           7690680.0   

               E_EXTRACTION_20  O_OTHER_20  X_ADJUSTMENTS_20  \
City                                                           
American Fork          80000.0  12896236.0         1169425.0   
Beaver City                0.0   1165909.0          218144.0   
Blanding                1000.0    697047.0          225818.0   
Bountiful              40000.0   6665067.0         3441209.0   
Brigham                 6000.0   3747382.0         1276922.0   

               R_AUTO_PRIVATE_20  S_PROF_TECH_SRV_20  G_GOVERNMENT_20  \
City                                                                    
American Fork          5500000.0           4721426.0          45000.0   
Beaver City             500000.0            141782.0              0.0   
Blanding                900000.0            139751.0         400000.0   
Bountiful              6000000.0           3618011.0         800000.0   
Brigham                3250000.0            764659.0           6000.0   

               S_RENTAL_LEASING_20  R_BUILDING_SUPPLY_20  R_CLOTHING_20  \
City                                                                      
American Fork            4239640.0            24434022.0      8831414.0   
Beaver City               161945.0              958291.0       126941.0   
Blanding                   59602.0             2580370.0       108254.0   
Bountiful                2180497.0             1557832.0      3394995.0   
Brigham                  1905406.0             1876175.0       534765.0   

               R_ELECTRONICS_20  R_GROCERY_BEV_20  R_FURNITURE_20  \
City                                                                
American Fork         6888175.0         4204031.0       4882872.0   
Beaver City             30764.0         2250000.0          3577.0   
Blanding               107663.0         2000000.0         10512.0   
Bountiful             4258239.0        22292365.0       1913518.0   
Brigham               1315595.0        11700758.0        535790.0   

               R_GAS_STATIONS_20  R_GE

### Load Q3 2019 sales tax revenue data

In [8]:
# load excel sheet into a dataframe
xlsx = '.\\Inputs\\2019-03-quarterly-sales.xlsx'
q3_2019 = pd.read_excel(xlsx, sheet_name='Table 9', header=5)

# Drop last empty two rows
q3_2019.drop(q3_2019.tail(2).index,inplace=True)

# forward fill values from merged cells
q3_2019['County'] = q3_2019['County'].fillna(method='ffill')
q3_2019['Location Code'] = q3_2019['Location Code'].fillna(method='ffill')
q3_2019['City'] = q3_2019['City'].fillna(method='ffill')

# Figures with less than 10 taxpayers have been rounded up per Tax Commission disclosure rules.
# Rename column and convert values to boolean
q3_2019.rename(columns={'Unnamed: 5':'Rounded Up'}, inplace=True)
crosswalk = {'*':True, np.nan:False} # County_ID
q3_2019['Rounded Up'] = q3_2019['Rounded Up'].map(crosswalk)

q3_2019.tail()

County  Location Code         City  \
2408  Weber County        29040.0  South Ogden   
2409  Weber County        29040.0  South Ogden   
2410  Weber County        29040.0  South Ogden   
2411  Weber County        29040.0  South Ogden   
2412  Weber County        29040.0  South Ogden   

                          Economic Sector (NAICS Code)     2019Q3  Rounded Up  
2408             FOOD SERVICES & DRINKING PLACES (722)  9854548.0       False  
2409  OTHER SERVICES-EXCEPT PUBLIC ADMINISTRATION (81)  1663672.0       False  
2410                        PUBLIC ADMINISTRATION (92)     7000.0        True  
2411                       PRIVATE MOTOR VEHICLE SALES  1250000.0        True  
2412                   PRIOR-PERIOD PAYMENTS & REFUNDS   532896.0       False

In [9]:
# pivot the table to get sectors as columns with tax values by city
sectors_q3_2019 = pd.pivot_table(q3_2019,values = '2019Q3',index ='City', columns = 'Economic Sector (NAICS Code)', 
                                 aggfunc='first')

# fill NAs with 0
sectors_q3_2019 = sectors_q3_2019.fillna(0)

# fix column names
#new_names = [replace(item) for item in list(sectors_q3_2019.columns)]
new_names = ['L_ACCOMMODATION_19', 'S_ADMIN_SUPPORT_19','A_AG_WILDLIFE_19','L_CULTURAL_REC_19','C_CONSTRUCTION_19',
'E_EDUCATION_19','S_FINANCIAL_19','L_RSTRNT_BAR_19','H_HEALTH_CARE_19','S_IT_DATA_19','S_CORPORATE_MGMT_19',
'M_MANUFACTURING_19','E_EXTRACTION_19','O_OTHER_19','X_ADJUSTMENTS_19','R_AUTO_PRIVATE_19','S_PROF_TECH_SRV_19',
'G_GOVERNMENT_19','S_RENTAL_LEASING_19','R_BUILDING_SUPPLY_19','R_CLOTHING_19','R_ELECTRONICS_19','R_GROCERY_BEV_19',
'R_FURNITURE_19','R_GAS_STATIONS_19','R_GENERAL_RETAIL_19','R_HEALTH_RETAIL_19','R_OTHER_RETAIL_19','R_AUTO_RETAIL_19',
'R_NONSTORE_RETAIL_19','R_SPORT_HOBBY_19','L_SPECIAL_EVENT_19','W_DISTRIBUTION_19','X_UNKNOWN_19','U_UTILITIES_19',
'W_WHLSALE_DURABLE_19','S_WHLSLE_ETRADE_19','W_WHLSLE_NDURABLE_19']
sectors_q3_2019.columns = new_names

# drop unknown column since its not present in 2020
sectors_q3_2019.drop('X_UNKNOWN_19', axis=1, inplace=True)
new_names.remove('X_UNKNOWN_19')

# get the total sales tax revenue
sectors_q3_2019['TOTAL_19'] = sectors_q3_2019[new_names].sum(axis=1)

sectors_q3_2019.head()

L_ACCOMMODATION_19  S_ADMIN_SUPPORT_19  A_AG_WILDLIFE_19  \
City                                                                      
American Fork            800000.0            668822.0          350000.0   
Beaver City             1486864.0             27631.0           60000.0   
Blanding                1500000.0             61062.0            1000.0   
Bountiful                200000.0           2313211.0            2000.0   
Brigham                 2000000.0            172922.0            2000.0   

               L_CULTURAL_REC_19  C_CONSTRUCTION_19  E_EDUCATION_19  \
City                                                                  
American Fork          1644861.0          8342718.0        169050.0   
Beaver City               8000.0           200000.0         20000.0   
Blanding                 60000.0           203713.0        150000.0   
Bountiful              1042613.0          1657783.0        144399.0   
Brigham                 800000.0          1482857.0         60540.0   

               S_FINANCIAL_19  L_RSTRNT_BAR_19  H_HEALTH_CARE_19  \
City                                                               
American Fork       1537569.0       24349796.0          388483.0   
Beaver City          256761.0        2805382.0           50000.0   
Blanding              17946.0         700000.0           70000.0   
Bountiful           1224894.0       13309802.0         1089477.0   
Brigham              203754.0        7103067.0          509334.0   

               S_IT_DATA_19  S_CORPORATE_MGMT_19  M_MANUFACTURING_19  \
City                                                                   
American Fork     9566264.0              15000.0           2918859.0   
Beaver City        440226.0               2000.0           1582128.0   
Blanding           423908.0               1000.0            221758.0   
Bountiful         4928417.0              35000.0           2503147.0   
Brigham           2728484.0               4000.0           8387846.0   

               E_EXTRACTION_19  O_OTHER_19  X_ADJUSTMENTS_19  \
City                                                           
American Fork          25000.0  11162595.0         2294792.0   
Beaver City                0.0   1009688.0          222285.0   
Blanding                   0.0    283537.0          214830.0   
Bountiful              30000.0   6902004.0         1174422.0   
Brigham                10000.0   3116643.0         1860373.0   

               R_AUTO_PRIVATE_19  S_PROF_TECH_SRV_19  G_GOVERNMENT_19  \
City                                                                    
American Fork          3250000.0           8069449.0          35000.0   
Beaver City             400000.0            114258.0              0.0   
Blanding                900000.0             71006.0         200000.0   
Bountiful              4750000.0           2313666.0         700000.0   
Brigham                2750000.0            452784.0           6000.0   

               S_RENTAL_LEASING_19  R_BUILDING_SUPPLY_19  R_CLOTHING_19  \
City                                                                      
American Fork            4118002.0            19753584.0      8371989.0   
Beaver City               101136.0              702714.0        81107.0   
Blanding                   93547.0             1668222.0        72090.0   
Bountiful                2361609.0             1155702.0      3015615.0   
Brigham                  1683923.0             1607103.0       434641.0   

               R_ELECTRONICS_19  R_GROCERY_BEV_19  R_FURNITURE_19  \
City                                                                
American Fork         7249934.0         4203317.0       3780567.0   
Beaver City             29373.0         2000000.0          3000.0   
Blanding               119245.0         2000000.0         12886.0   
Bountiful             3470102.0        20434125.0       1705584.0   
Brigham               1100882.0        10088650.0        510411.0   

               R_GAS_STATIONS_19  R_GE

### Calculate the differences 
*2020 - 2019 = Diff*

In [10]:
# Check join output shape
print(sectors_q3_2019.shape)
print(sectors_q3_2020.shape)

(66, 38)
(66, 38)


In [11]:
# merge the 2019-2020 difference with the sdf
sectors_q3_diff = pd.DataFrame(sectors_q3_2020.values - sectors_q3_2019.values)

# add back the column names
new_names = ['L_ACCOMMODATION_D20', 'S_ADMIN_SUPPORT_D20','A_AG_WILDLIFE_D20','L_CULTURAL_REC_D20','C_CONSTRUCTION_D20',
'E_EDUCATION_D20','S_FINANCIAL_D20','L_RSTRNT_BAR_D20','H_HEALTH_CARE_D20','S_IT_DATA_D20','S_CORPORATE_MGMT_D20',
'M_MANUFACTURING_D20','E_EXTRACTION_D20','O_OTHER_D20','X_ADJUSTMENTS_D20','R_AUTO_PRIVATE_D20','S_PROF_TECH_SRV_D20',
'G_GOVERNMENT_D20','S_RENTAL_LEASING_D20','R_BUILDING_SUPPLY_D20','R_CLOTHING_D20','R_ELECTRONICS_D20','R_GROCERY_BEV_D20',
'R_FURNITURE_D20','R_GAS_STATIONS_D20','R_GENERAL_RETAIL_D20','R_HEALTH_RETAIL_D20','R_OTHER_RETAIL_D20','R_AUTO_RETAIL_D20',
'R_NONSTORE_RETAIL_D20','R_SPORT_HOBBY_D20','L_SPECIAL_EVENT_D20','W_DISTRIBUTION_D20','U_UTILITIES_D20',
'W_WHLSALE_DURABLE_D20','S_WHLSLE_ETRADE_D20','W_WHLSLE_NDURABLE_D20', 'TOTAL_D20']

sectors_q3_diff.columns = new_names

# add back cities as index
sectors_q3_diff['City'] = sectors_q3_2020.index
sectors_q3_diff = sectors_q3_diff.set_index('City')

#check table
sectors_q3_diff.head(10)

L_ACCOMMODATION_D20  S_ADMIN_SUPPORT_D20  \
City                                                           
American Fork                 -189390.0              15568.0   
Beaver City                    309836.0              -5494.0   
Blanding                      -858412.0               2730.0   
Bountiful                           0.0            -784974.0   
Brigham                       -762732.0              25174.0   
Cedar City                    -977142.0             672870.0   
Centerville                     15000.0              93924.0   
Clearfield                          0.0             105639.0   
Cottonwood Heights              27834.0            -155178.0   
Delta                               0.0              -6223.0   

                    A_AG_WILDLIFE_D20  L_CULTURAL_REC_D20  C_CONSTRUCTION_D20  \
City                                                                            
American Fork                350000.0             97732.0            707985.0   
Beaver City                   20000.0             -1000.0            -50000.0   
Blanding                          0.0            -20000.0            -53713.0   
Bountiful                       971.0             55788.0            318638.0   
Brigham                        1000.0           -100000.0           -871095.0   
Cedar City                     1831.0              5509.0           1230319.0   
Centerville                       0.0             12000.0           -272030.0   
Clearfield                     2000.0           -230044.0            168460.0   
Cottonwood Heights             1000.0          -1720000.0            994406.0   
Delta                             0.0                 0.0           1574415.0   

                    E_EDUCATION_D20  S_FINANCIAL_D20  L_RSTRNT_BAR_D20  \
City                                                                     
American Fork               27222.0         782550.0         3108914.0   
Beaver City                -14000.0         231085.0          401758.0   
Blanding                        0.0            485.0         -250000.0   
Bountiful                    -554.0          38148.0         -287952.0   
Brigham                     -6274.0          90024.0          486466.0   
Cedar City                -410014.0        -613688.0         2201010.0   
Centerville                  8549.0         -78639.0          297630.0   
Clearfield                  22699.0           9250.0          679235.0   
Cottonwood Heights          48169.0        -948979.0        -1771987.0   
Delta                           0.0         -78716.0          106977.0   

                    H_HEALTH_CARE_D20  S_IT_DATA_D20  S_CORPORATE_MGMT_D20  \
City                                                                         
American Fork                 21403.0     -2274452.0              -13000.0   
Beaver City                   10000.0        29947.0               -1000.0   
Blanding                     -30000.0        30471.0                2000.0   
Bountiful                    140190.0       659866.0              -10000.0   
Brigham                     -159537.0       -70446.0                4000.0   
Cedar City                    -3308.0        86260.0                5000.0   
Centerville                    8725.0     -1379614.0              -45000.0   
Clearfield                   -11354.0       599259.0              250000.0   
Cottonwood Heights           -39100.0       885814.0             -200000.0   
Delta                         40000.0         9615.0               -5000.0   

                    M_MANUFACTURING_D20  E_EXTRACTION_D20  O_OTHER_D20  \
City                                                                     
American Fork                  452281.0           55000.0    1733641.0   
Beaver City                    298839.0               0.0     156221.0   
Blanding                       -93674.0            1000.0     413510.0   
Bountiful                       76311.0           10000.0    -236937.0   
Brigham 

### Load municipalities and townships shapefile and join to formatted sales tax data

In [12]:
# load cities shapefile into pandas spatial dataframe
cities_shp = '.\\Inputs\\Cities.shp'
cities_sdf = pd.DataFrame.spatial.from_featureclass(cities_shp)

# Rename columns
cities_sdf.columns = ['FID', 'NAME', 'SOURCE', 'SALESTAXID', 'POPLASTCEN', 'POPLASTEST',
       'AREA_SQMI', 'SHAPE']

In [13]:
# merge the 2020 sdf with the pivoted data
cities_sdf2 = cities_sdf.merge(sectors_q3_2019, left_on= "NAME", right_on="City", how="inner")
cities_sdf2 = cities_sdf2.merge(sectors_q3_2020, left_on= "NAME", right_on="City", how="inner")
cities_sdf2 = cities_sdf2.merge(sectors_q3_diff, left_on= "NAME", right_on="City", how="inner")
cities_sdf2.tail(10)


FID              NAME  SOURCE SALESTAXID  POPLASTCEN  POPLASTEST  \
55  222          Syracuse  Cities        049       25113       30400   
56  224      Taylorsville  Cities        142       59989       60192   
57  225       Tooele City  Cities        048       32629       35251   
58  228         Tremonton  Cities        113        7828        8882   
59  232            Vernal  Cities        024        9460       10370   
60  238   Washington City  Cities        027       19640       27686   
61  243    West Bountiful  Cities        061        5420        5731   
62  245       West Jordan  Cities        155      105891      116046   
63  247  West Valley City  Cities        167      133796      136401   
64  252       Woods Cross  Cities        057       10086       11328   

    AREA_SQMI                                              SHAPE  \
55  10.193900  {'rings': [[[411406.60675760085, 4549994.30436...   
56  10.839700  {'rings': [[[422191.826826254, 4504303.3642982...   
57  24.087099  {'rings': [[[392501.67668449413, 4486111.93423...   
58   7.955100  {'rings': [[[399303.9767575987, 4621607.124490...   
59   4.632710  {'rings': [[[627188.8071560834, 4479658.624106...   
60  36.165600  {'rings': [[[278123.416362615, 4102773.6636791...   
61   3.299440  {'rings': [[[421904.10680578306, 4528452.80425...   
62  32.306999  {'rings': [[[415038.796736156, 4499131.1642154...   
63  35.799702  {'rings': [[[418708.9467166035, 4508796.864202...   
64   3.835550  {'rings': [[[424779.70679974474, 4524289.10421...   

    L_ACCOMMODATION_19  S_ADMIN_SUPPORT_19  A_AG_WILDLIFE_19  \
55             45000.0             72936.0            1000.0   
56            450000.0            481676.0            9000.0   
57           1750000.0            190433.0            3000.0   
58           1250000.0           1038981.0            5000.0   
59           3467889.0            171297.0            3000.0   
60           2098062.0            232774.0          150000.0   
61            800000.0             33602.0           35000.0   
62           2250000.0           1529560.0          350000.0   
63           9059280.0           4328340.0           40000.0   
64           1750000.0            395073.0            1000.0   

    L_CULTURAL_REC_19  C_CONSTRUCTION_19  E_EDUCATION_19  S_FINANCIAL_19  \
55          1250000.0           223496.0         28558.0        554255.0   
56          2042324.0          1613198.0        131432.0       2204654.0   
57          1250000.0          1324662.0         58031.0        926712.0   
58           450000.0          1005745.0          6000.0        104887.0   
59           300000.0           346649.0         32811.0        467390.0   
60          1000000.0          1016778.0          6523.0        450852.0   
61           400000.0           350262.0          5000.0        127283.0   
62          2241101.0         11246026.0        193131.0       2424064.0   
63          4895306.0         29100683.0        235605.0       3444233.0   
64             1000.0          1028462.0         80865.0        254204.0   

    L_RSTRNT_BAR_19  H_HEALTH_CARE_19  S_IT_DATA_19  S_CORPORATE_MGMT_19  \
55        4907056.0           70000.0     2700893.0              20000.0   
56       28130157.0          110489.0     8563764.0              35000.0   
57       12654430.0          896617.0     4492589.0              20000.0   
58        3177843.0            2000.0      959894.0               3000.0   
59       10929856.0          331909.0     4741490.0               4000.0   
60        9630592.0            5000.0     2839398.0               9000.0   
61        3892466.0            5000.0     1627256.0               3000.0   
62       37394254.0          998973.0    21446770.0             200000.0   
63       57423994.0          968752.0    24403526.0               9000.0   
64        5060453.0           30000.0     2447935.0               2000.0   

    M_MANUFACTURING_19  E_EXTRACTION_19  O_OTHER_19  X_ADJUSTMENTS_19  \
55        

In [14]:
# Confirm join count, Salt Lake County (Unincorporated) should be the only one not joined
l1 = list(q3_2020['City'].value_counts().index)
l2 = list(cities_sdf['NAME'].value_counts().index)
list(set(l1) - set(l2))

['Salt Lake County (Unincorporated)']

In [15]:
# Create file gdb and export sdf to feature class
outputs = '.\\Outputs'

gdb = os.path.join(outputs, "taxable_sales.gdb")
if not arcpy.Exists(gdb):
    arcpy.CreateFileGDB_management(outputs, "taxable_sales.gdb")

cities_sdf2.spatial.to_featureclass(location=os.path.join(outputs, "taxable_sales.gdb","taxable_sales_utah_q3"))

'E:\\Projects\\UT_Sales_Tax_Analysis\\Outputs\\taxable_sales.gdb\\taxable_sales_utah_q3'

### Analysis

In [16]:
# Get year specific column names
standard_cols = ['fid','name','source','salestaxid','poplastcen','poplastcen','area_sqmi', 'SHAPE']
names_19 = standard_cols + [col for col in cities_sdf2.columns if '_19' in col]
names_20 = standard_cols + [col for col in cities_sdf2.columns if '_20' in col]
names_D20 = standard_cols + [col for col in cities_sdf2.columns if '_d20' in col]

# subset columns by time
cities_q3_19 = cities_sdf2[names_19]
cities_q3_20 = cities_sdf2[names_20]
cities_q3_d20 = cities_sdf2[names_D20]


In [17]:
# 10 cities with smallest difference of revenue
cities_q3_d20[['name', 'poplastcen', 'area_sqmi', 'total_d20']].sort_values('total_d20', ascending=False).head(10)

name  poplastcen  poplastcen  area_sqmi    total_d20
54        St George       76548       76548  77.209900  132030878.0
62      West Jordan      105891      105891  32.306999   90642198.0
34             Orem       92993       92993  18.628000   67079183.0
33            Ogden       85183       85183  27.493500   63764733.0
23            Logan       49612       49612  18.485800   63421246.0
48     South Jordan       51057       51057  22.263201   62723200.0
20           Layton       69457       69457  22.576200   61634561.0
51     Spanish Fork       36232       36232  16.249701   52951283.0
21             Lehi       49315       49315  28.898199   51286416.0
50  South Salt Lake       24282       24282   6.936550   51193410.0

In [18]:
# 10 cities with largest difference of revenue
cities_q3_d20[['name', 'poplastcen', 'area_sqmi', 'total_d20']].sort_values('total_d20', ascending=True).head(10)

name  poplastcen  poplastcen   area_sqmi    total_d20
45   Salt Lake City      190749      190749  110.734001 -139795789.0
12       Farmington       18752       18752   10.035100  -27566629.0
35        Park City        7709        7709   20.341700  -10722173.0
43        Roosevelt        6260        6260    6.757070  -10027223.0
25          Midvale       28644       28644    5.853220   -7935591.0
28       Monticello        2053        2053    3.349260   -1426458.0
32  North Salt Lake       16930       16930    8.512150    -431477.0
17            Kanab        4413        4413   14.473200     160837.0
59           Vernal        9460        9460    4.632710    1026651.0
2          Blanding        3514        3514   13.205500    1322505.0

#### Some industries of interest:
- RETAIL-GASOLINE STATIONS (447)
- RETAIL-FOOD & BEVERAGE STORES (445)
- REAL ESTATE, RENTAL & LEASING (53)
- ARTS, ENTERTAINMENT AND RECREATION (71)
- ACCOMMODATION (721)